tools/regtesting/do_regtest.py

```diff
@@ -0,0 +1,437 @@
+    num_restrictdirs = 0
+    for batch in batches:
+        if not batch.requirements_satisfied(flags, cfg.mpiranks):
+            print(f"Skipping {batch.name} because its requirements are not satisfied.")
+        elif not any([re.match(p, batch.name) for p in cfg.restrictdirs]):
+            num_restrictdirs += 1
+        elif batch.name in ALLWAYS_SKIP_DIRS:
+            print(f"Skipping {batch.name} because it doesn't work with the cp2k shell.")
+        else:
+            tasks.append(asyncio.create_task(process_batch(batch, cfg)))  # launch
+
+    if num_restrictdirs:
+        print(f"Skipping {num_restrictdirs} test directories because of --restrictdir.")
+    if not tasks:
+        print("\nNot test directories selected, check --restrictdir filter.")
+        sys.exit(1)
+
+    # Wait for tasks to finish and print their results.
+    print(f"Launched {len(tasks)} test directories and {cfg.num_workers} worker...\n")
+    all_results: List[TestResult] = []
+    with open(cfg.error_summary, "wt", encoding="utf8", errors="replace") as err_fh:
+        for num_done, task in enumerate(asyncio.as_completed(tasks)):
+            batch_result = await task
+            all_results += batch_result.results
+            print(f">>> {batch_result.batch.workdir}")
+            print("\n".join([str(r) for r in batch_result.results]))
+            print(f"<<< {batch_result.batch.workdir} ({num_done + 1}", end="")
+            print(f" of {len(tasks)}) done in {batch_result.duration:.2f} sec")
+            sys.stdout.flush()
+            err_fh.write("\n".join([r.error for r in batch_result.results if r.error]))
+            err_fh.flush()
+
+    print("\n--------------------------------- Errors -------------------------------")
+    print("\n".join([r.error for r in all_results if r.error]))
+
+    print("\n-------------------------------- Timings -------------------------------")
+    timings = sorted([r.duration for r in all_results])
+    print('Plot: name="timings", title="Timing Distribution", ylabel="time [s]"')
+    for p in (100, 99, 98, 95, 90, 80):
+        v = percentile(timings, p / 100.0)
+        print(f'PlotPoint: name="{p}th_percentile", plot="timings", ', end="")
+        print(f'label="{p}th %ile", y={v:.2f}, yerr=0.0')
+
+    print("\n-------------------------------- Summary -------------------------------")
+    total_duration = time.perf_counter() - start_time
+    num_tests = len(all_results)
+    num_failed = sum([r.status in ("TIMEOUT", "RUNTIME FAIL") for r in all_results])
+    num_wrong = sum([r.status == "WRONG RESULT" for r in all_results])
+    num_ok = sum([r.status == "OK" for r in all_results])
+    print(f"Number of FAILED  tests {num_failed}")
+    print(f"Number of WRONG   tests {num_wrong}")
+    print(f"Number of CORRECT tests {num_ok}")
+    print(f"Total number of   tests {num_tests}")
+    summary = f"\nSummary: correct: {num_ok} / {num_tests}"
+    summary += f"; wrong: {num_wrong}" if num_wrong > 0 else ""
```

Comment:
```suggestion
    summary += f"; wrong: {num_wrong if num_wrong > 0 else ''}"
```

... and then thoe whole thing can be one string:

```
summary = (
    f"\nSummary: correct: {num_ok} / {num_tests}"
    f"; wrong: {num_wrong if num_wrong > 0 else ''}"
    f"; failed: {num_failed  if num_failed > 0 else ''}"
    f"; {total_duration/60.0:.0f}min"
    )

Reply:
I don't think this will work because I also want `"; failed: "` to disappear when there are no failures.

Reply:
ok, I see, but then I'd probably use  a prefixed `if` for clearness rather than appending an empty string:

```python
if num_failed:
    summary += f"..."
```

Reply:
If this were some serious business logic then I would agree. But just for assembling this string I'd rather not spend the extra lines.

Reply:
it makes code more readable